lib/srv/db/cloud/watchers/azure.go

```diff
@@ -0,0 +1,186 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package watchers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/gravitational/teleport/api/types"
+	apiutils "github.com/gravitational/teleport/api/utils"
+	"github.com/gravitational/teleport/lib/cloud/azure"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/srv/db/common"
+	"github.com/gravitational/teleport/lib/utils"
+
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+)
+
+// newAzureFetcher returns a Azure DB server fetcher for the provided subscription, group, regions, and tags.
+func newAzureFetcher(config azureFetcherConfig) (*azureFetcher, error) {
+	if err := config.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	fetcher := &azureFetcher{
+		cfg: config,
+		log: logrus.WithFields(logrus.Fields{
+			trace.Component: "watch:azure",
+			"labels":        config.Labels,
+			"regions":       config.Regions,
+			"group":         config.ResourceGroup,
+			"subscription":  config.Subscription,
+		}),
+	}
+	return fetcher, nil
+}
+
+// azureFetcherConfig is the Azure database servers fetcher configuration.
+type azureFetcherConfig struct {
+	// Client is the Azure API client.
+	Client azure.DBServersClient
+	// Subscription is the Azure subscription being fetched from by the fetcher.
+	Subscription string
+	// ResourceGroup is a selector to match cloud resource group.
+	ResourceGroup string
+	// Labels is a selector to match cloud databases.
+	Labels types.Labels
+	// Regions is the Azure regions selectors to match cloud databases.
+	Regions []string
+	// regionMatches returns whether a given region matches the configured Regions selector
+	regionMatches func(region string) bool
+}
+
+// CheckAndSetDefaults validates the config and sets defaults.
+func (c *azureFetcherConfig) CheckAndSetDefaults() error {
+	if c.Client == nil {
+		return trace.BadParameter("missing parameter Client")
+	}
+	if len(c.Subscription) == 0 {
+		return trace.BadParameter("missing parameter Subscription")
+	}
+	if len(c.ResourceGroup) == 0 {
+		return trace.BadParameter("missing parameter ResourceGroup")
+	}
+	if len(c.Labels) == 0 {
+		return trace.BadParameter("missing parameter Labels")
+	}
+	if len(c.Regions) == 0 {
+		return trace.BadParameter("missing parameter Regions")
+	}
+	if apiutils.SliceContainsStr(c.Regions, types.Wildcard) {
+		// wildcard matches all regions
+		c.regionMatches = func(_ string) bool { return true }
+	} else {
+		regionSet := utils.StringsSet(c.Regions)
+		c.regionMatches = func(region string) bool {
+			_, ok := regionSet[region]
+			return ok
+		}
+	}
```

Comment:
nit: 
Can we move this into a separate function and extract the filtering logic from `CheckAndSetDefaults`  function ?

Reply:
It was an optimization to avoid checking for the wildcard in the slice, and to avoid constructing regionSet each time.

Reply:
IMHO, we should not have one fetcher per region, since the client cannot filter by region anyway. If we pass the entire region list to a fetcher, it also reduces number of the `List` calls if there is more than one region specified.

Reply:
I changed this